File: src/harness/memory.py
from __future__ import annotations

from typing import Any, Dict, List

from .types import TurnRecord


class MemoryManager:
    """Manages the working memory window sent to the LLM each turn.

    Maintains a rolling window of recent turns and compresses older history
    into a summary string that preserves tagged observations (``constraint:``,
    ``todo:``, ``evidence:``).

    Args:
        max_history_turns: Number of most-recent turns included verbatim in
            the working memory dict.
    """

    def __init__(self, max_history_turns: int = 8) -> None:
        self.max_history_turns = max_history_turns
        self.summary: str = ""

    def build_working_memory(
        self, goal: str, context: Dict[str, Any], turns: List[TurnRecord]
    ) -> Dict[str, Any]:
        """Build the working memory dict for the current turn.

        Args:
            goal: The task goal string.
            context: Extra context key/value pairs.
            turns: Full turn history so far.

        Returns:
            A dict with keys ``goal``, ``context``, ``summary_memory``, and
            ``history`` (the most recent ``max_history_turns`` turns).
        """
        recent_turns = turns[-self.max_history_turns :]
        history = [
            {
                "turn": t.turn,
                "action": t.llm_action,
                "observation": t.observation,
                "tool_result": t.tool_result,
            }
            for t in recent_turns
        ]

        return {
            "goal": goal,
            "context": context,
            "summary_memory": self.summary,
            "history": history,
        }

    def maybe_compress(self, turns: List[TurnRecord], max_total_turns: int = 12) -> bool:
        if len(turns) <= max_total_turns:
            return False

        old_turns = turns[: -self.max_history_turns]
        if not old_turns:
            return False

        constraints = self._extract_tagged_observations(old_turns, "constraint:")
        open_items = self._extract_tagged_observations(old_turns, "todo:")
        evidence = self._extract_tagged_observations(old_turns, "evidence:")
        brief_lines = [f"turn {t.turn}: {t.observation}" for t in old_turns[-4:]]

        summary_parts = []
        if constraints:
            summary_parts.append(f"Constraints: {'; '.join(constraints)}")
        if open_items:
            summary_parts.append(f"Open items: {'; '.join(open_items)}")
        if evidence:
            summary_parts.append(f"Evidence: {'; '.join(evidence)}")
        if brief_lines:
            summary_parts.append(f"Recent compressed history: {' | '.join(brief_lines)}")

        self.summary = " || ".join(summary_parts)
        return True

    def summarize_run(self, goal: str, turns: List[TurnRecord], stop_reason: str) -> str:
        """Return a compact one-line summary of a completed run.

        Suitable for accumulating cross-goal context in a persistent session.

        Args:
            goal: The task goal string.
            turns: Full turn list from the completed run.
            stop_reason: The stop reason from :class:`~harness.types.RunResult`.

        Returns:
            A short summary string with goal, stop reason, turn count, and any
            tagged observations.
        """
        constraints = self._extract_tagged_observations(turns, "constraint:")
        evidence = self._extract_tagged_observations(turns, "evidence:")

        parts = [f"[Goal: {goal[:80]}] stop={stop_reason} turns={len(turns)}"]
        if constraints:
            parts.append(f"constraints: {'; '.join(constraints[:3])}")
        if evidence:
            parts.append(f"evidence: {'; '.join(evidence[:3])}")

        return " | ".join(parts)

    @staticmethod
    def _extract_tagged_observations(turns: List[TurnRecord], prefix: str) -> List[str]:
        seen: set[str] = set()
        values: List[str] = []
        for turn in turns:
            observation = turn.observation.strip()
            if observation.lower().startswith(prefix):
                value = observation[len(prefix) :].strip()
                if value and value not in seen:
                    seen.add(value)
                    values.append(value)
        return values

File: ts/src/memory.ts
/**
 * Memory management — rolling window + compression + tagged observation extraction.
 */

import type { TurnRecord } from "./types.js";

export const MAX_OBSERVATION_CHARS = 2000;

// --- Public interfaces ---

export interface HistoryEntry {
  turn: number;
  action: Record<string, unknown>;
  observation: string;
  tool_result: Record<string, unknown> | null;
}

export interface WorkingMemory {
  goal: string;
  context: Record<string, unknown>;
  summary_memory: string;
  history: HistoryEntry[];
}

// --- Helper functions (exported for testing) ---

/** Truncate `text` to `limit` chars, appending a marker if cut. */
export function truncateStr(text: string, limit: number): string {
  if (text.length <= limit) return text;
  return text.slice(0, limit) + `\n[observation truncated at ${limit} chars]`;
}

/** Return a shallow copy of `result` with long string fields truncated. */
export function truncateToolResult(result: unknown): unknown {
  if (result === null || typeof result !== "object" || Array.isArray(result)) {
    return result;
  }

  const copy = { ...(result as Record<string, unknown>) };
  const output = copy.output;

  if (output !== null && typeof output === "object" && !Array.isArray(output)) {
    const outputObj = output as Record<string, unknown>;
    if ("content" in outputObj && typeof outputObj.content === "string") {
      if (outputObj.content.length > MAX_OBSERVATION_CHARS) {
        const outputCopy = { ...outputObj };
        outputCopy.content =
          outputObj.content.slice(0, MAX_OBSERVATION_CHARS) +
          `\n[observation truncated at ${MAX_OBSERVATION_CHARS} chars]`;
        copy.output = outputCopy;
      }
    }
  } else if (typeof output === "string" && output.length > MAX_OBSERVATION_CHARS) {
    copy.output =
      output.slice(0, MAX_OBSERVATION_CHARS) +
      `\n[observation truncated at ${MAX_OBSERVATION_CHARS} chars]`;
  }

  return copy;
}

/** Extract lines starting with `prefix` (case-insensitive) from turn observations. Deduped. */
function extractTaggedObservations(turns: TurnRecord[], prefix: string): string[] {
  const seen = new Set<string>();
  const values: string[] = [];
  for (const turn of turns) {
    const observation = turn.observation.trim();
    if (observation.toLowerCase().startsWith(prefix.toLowerCase())) {
      const value = observation.slice(prefix.length).trim();
      if (value && !seen.has(value)) {
        seen.add(value);
        values.push(value);
      }
    }
  }
  return values;
}

// --- MemoryManager class ---

export class MemoryManager {
  public summary = "";

  constructor(public readonly maxHistoryTurns: number = 8) {}

  /** Build the working memory dict for the current turn. */
  buildWorkingMemory(
    goal: string,
    context: Record<string, unknown>,
    turns: TurnRecord[],
  ): WorkingMemory {
    const recentTurns = turns.slice(-this.maxHistoryTurns);
    const history: HistoryEntry[] = recentTurns.map((t) => ({
      turn: t.turn,
      action: t.llm_action,
      observation: truncateStr(t.observation, MAX_OBSERVATION_CHARS),
      tool_result: truncateToolResult(t.tool_result) as Record<string, unknown> | null,
    }));

    return {
      goal,
      context,
      summary_memory: this.summary,
      history,
    };
  }

  /** Compress old turns into summary when history exceeds threshold. Returns true if compressed. */
  maybeCompress(turns: TurnRecord[], maxTotalTurns?: number): boolean {
    const threshold = maxTotalTurns ?? this.maxHistoryTurns + 4;
    if (turns.length <= threshold) return false;

    const oldTurns = turns.slice(0, -this.maxHistoryTurns);
    if (oldTurns.length === 0) return false;

    const constraints = extractTaggedObservations(oldTurns, "constraint:");
    const openItems = extractTaggedObservations(oldTurns, "todo:");
    const evidence = extractTaggedObservations(oldTurns, "evidence:");
    const briefLines = oldTurns
      .slice(-4)
      .map((t) => `turn ${t.turn}: ${t.observation}`);

    const parts: string[] = [];
    if (constraints.length > 0) parts.push(`Constraints: ${constraints.join("; ")}`);
    if (openItems.length > 0) parts.push(`Open items: ${openItems.join("; ")}`);
    if (evidence.length > 0) parts.push(`Evidence: ${evidence.join("; ")}`);
    if (briefLines.length > 0) parts.push(`Recent compressed history: ${briefLines.join(" | ")}`);

    this.summary = parts.join(" || ");
    return true;
  }

  /** Return a compact one-line summary of a completed run. */
  summarizeRun(goal: string, turns: TurnRecord[], stopReason: string): string {
    const constraints = extractTaggedObservations(turns, "constraint:");
    const evidence = extractTaggedObservations(turns, "evidence:");

    const parts = [`[Goal: ${goal.slice(0, 80)}] stop=${stopReason} turns=${turns.length}`];
    if (constraints.length > 0) parts.push(`constraints: ${constraints.slice(0, 3).join("; ")}`);
    if (evidence.length > 0) parts.push(`evidence: ${evidence.slice(0, 3).join("; ")}`);

    return parts.join(" | ");
  }
}
